searchcore/src/vespa/searchcore/bmcluster/bm_feeder.cpp

```diff
@@ -0,0 +1,199 @@
+// Copyright Yahoo. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+
+#include "bm_feeder.h"
+#include "avg_sampler.h"
+#include "bm_feed_params.h"
+#include "bm_range.h"
+#include "bucket_selector.h"
+#include "pending_tracker.h"
+#include "i_bm_feed_handler.h"
+#include <vespa/document/base/documentid.h>
+#include <vespa/document/bucket/bucket.h>
+#include <vespa/document/fieldset/fieldsets.h>
+#include <vespa/document/fieldvalue/document.h>
+#include <vespa/document/repo/documenttyperepo.h>
+#include <vespa/document/test/make_bucket_space.h>
+#include <vespa/document/update/documentupdate.h>
+#include <vespa/vespalib/util/lambdatask.h>
+#include <vespa/vespalib/util/threadstackexecutor.h>
+#include <cassert>
+#include <chrono>
+
+#include <vespa/log/log.h>
+LOG_SETUP(".bmcluster.bm_feeder");
+
+using document::Document;
+using document::DocumentId;
+using document::DocumentTypeRepo;
+using document::DocumentUpdate;
+using vespalib::makeLambdaTask;
+
+namespace search::bmcluster {
+
+BmFeeder::BmFeeder(std::shared_ptr<const DocumentTypeRepo> repo, IBmFeedHandler& feed_handler, vespalib::ThreadStackExecutor& executor)
+    : _repo(std::move(repo)),
+      _bucket_space(document::test::makeBucketSpace("test")),
+      _feed_handler(feed_handler),
+      _executor(executor)
+{
+}
+
+BmFeeder::~BmFeeder() = default;
+
+void
+BmFeeder::put_async_task(uint32_t max_pending, BmRange range, const vespalib::nbostream &serialized_feed, int64_t time_bias)
```

Comment:
Idea for a future PR: There are a lot of similarities between the different functions in this class that feed data via the feed handler. Perhaps some refactoring can reduce code duplication.